core/file.go

```diff
@@ -156,57 +164,296 @@ func (file *File) Evaluate(ctx context.Context) (*buildkit.Result, error) {
 }
 
 // Contents handles file content retrieval
-func (file *File) Contents(ctx context.Context) ([]byte, error) {
-	query, err := CurrentQuery(ctx)
+func (file *File) Contents(ctx context.Context, offset, limit *int) ([]byte, error) {
+	if limit != nil && *limit == 0 {
+		// edge case: 0 limit, possibly from maths, just don't do anything
+		return nil, nil
+	}
+
+	r, err := file.ReadCloser(ctx)
 	if err != nil {
 		return nil, err
 	}
-	bk, err := query.Buildkit(ctx)
+	defer r.Close()
+
+	var buf bytes.Buffer
+	w := &limitedWriter{
+		Limit:  buildkit.MaxFileContentsSize,
+		Writer: &buf,
+	}
+
+	if offset != nil || limit != nil {
+		br := bufio.NewReader(r)
+		lineNum := 1
+		readLines := 0
+		for {
+			line, err := br.ReadBytes('\n')
+			if err != nil && err != io.EOF {
+				return nil, err
+			}
+
+			if offset == nil || lineNum > *offset {
+				w.Write(line)
+				readLines++
+				if limit != nil && readLines == *limit {
+					break
+				}
+			}
+
+			if err == io.EOF {
+				break
+			}
+
+			lineNum++
+		}
+	} else {
+		_, err := io.Copy(w, r)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return buf.Bytes(), nil
+}
+
+type limitedWriter struct {
+	Limit int
+	io.Writer
+	wrote int
+}
+
+func (cw *limitedWriter) Write(p []byte) (int, error) {
+	if cw.wrote+len(p) > cw.Limit {
+		return 0, fmt.Errorf("file size %d exceeds limit %d", cw.wrote+len(p), buildkit.MaxFileContentsSize)
+	}
+	n, err := cw.Writer.Write(p)
 	if err != nil {
-		return nil, fmt.Errorf("failed to get buildkit client: %w", err)
+		return n, err
 	}
+	cw.wrote += n
+	return n, nil
+}
 
-	ref, err := bkRef(ctx, bk, file.LLB)
+// Reader returns an io.Reader for the file contents
+func (file *File) ReadCloser(ctx context.Context) (io.ReadCloser, error) {
+	ref, err := getRefOrEvaluate(ctx, file)
+	if err != nil {
+		return nil, err
+	}
+
+	mountable, err := ref.Mount(ctx, true, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get mountable reference: %w", err)
+	}
+
+	lm := snapshot.LocalMounter(mountable)
+	defer lm.Unmount()
```

Comment:
We do this unmount unconditionally but also call Unmount in fileReader.Close, which feels off

Reply:
Woops yea that's janky. I guess it's been unmounting immediately, which works anyway since the file is already opened.

Just noticed this is completely redundant with `File.Open` though. That one only supports `LLB` but #10907 will make it work for `Result` too, so I'll just remove this.